### PR TITLE
Feature/trigonometric funcs

### PR DESCRIPTION
**Pull request by ManiacMaxo**

### CODE REVIEW

**Comment by ManiacMaxo** [comment]
I forgot unit test, don't review just yet

**Comment by ManiacMaxo** [comment]
@VIVelev merge when you approve

File: nujo/autodiff/_functions/_trigonometric.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+__all__ = ['_Sin', '_Cos', '_Tan']
+
+# ====================================================================================================
+
+
+class _Sin(Function):
+    ''' Differable sine function
+
+    sin(X) = (e^iX - e^-iX) / 2i
```
Comment (VIVelev) [inline nujo/autodiff/_functions/_trigonometric.py]: Why do you add this identity here, unnecessary.

File: nujo/autodiff/_functions/_trigonometric.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    '''
+    def __init__(self, input: Union[Tensor, ndarray, List[Number], Number]):
+        super(_Sin, self).__init__(input)
+
+    def forward(self) -> ndarray:
+        return sin(self.children[0].value)
+
+    def backward(self) -> Function.T:
```
Comment (VIVelev) [inline nujo/autodiff/_functions/_trigonometric.py]: the `backward` method signature is not like that, check develop

File: nujo/autodiff/_functions/_trigonometric.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        # cos(X)
+        return Tensor(_Cos(self.children[0].value)())
```
Comment (VIVelev) [inline nujo/autodiff/_functions/_trigonometric.py]: You should not use nujo Functions in other nujo Functions, just return numpy's cos(value)

File: nujo/autodiff/_functions/_trigonometric.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+
+# ====================================================================================================
+
+
+class _Cos(Function):
+    ''' Differable cosine function
+
+    cos(X) = (e^iX + e^-iX) / i
```
Comment (VIVelev) [inline nujo/autodiff/_functions/_trigonometric.py]: I don't even know why you have searched for an identity like this one :D

File: nujo/autodiff/_functions/_trigonometric.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+        return Tensor(-(_Sin(self.children[0].value)()))
+
+
+# ====================================================================================================
+
+
+class _Tan(Function):
+    ''' Differable tangent function
+
+    tan(X) = sin(X) / cos(X)
```
Comment (VIVelev) [inline nujo/autodiff/_functions/_trigonometric.py:53]: That's more like it

File: nujo/autodiff/_functions/_trigonometric.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    '''
+    def __init__(self, input: Union[Tensor, ndarray, List[Number], Number]):
+        super(_Cos, self).__init__(input)
+
+    def forward(self) -> ndarray:
+        return cos(self.children[0].value)
+
+    def backward(self) -> Function.T:
+        # -sin(X)
+        return Tensor(-(_Sin(self.children[0].value)()))
```
Comment (VIVelev) [inline nujo/autodiff/_functions/_trigonometric.py]: Same, `backward` signature - incorrect and don't use `_Sin` (nujo Function) here, use numpy's `sin`

File: nujo/autodiff/_functions/_trigonometric.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+    '''
+    def __init__(self, input: Union[Tensor, ndarray, List[Number], Number]):
+        super(_Tan, self).__init__(input)
+
+    def forward(self) -> ndarray:
+        return tan(self.children[0].value)
+
+    def backward(self) -> Function.T:
+        # sec^2(X)
+        return Tensor((1 / _Cos(self.children[0].value)())**2)
```
Comment (VIVelev) [inline nujo/autodiff/_functions/_trigonometric.py]: You know what to do

File: tests/test_autodiff/test_functions/test_trigonometric.py [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+    A = input
+    sine = funcs._Sin(A)
+
+    # Test Forwardprop
+    C = sine()
+    assert isinstance(C, Tensor)
+    assert allclose(sin(A), C.value)
+
+    # Test Backprop
+    grad_A = sine.backward()
```
Comment (VIVelev) [inline tests/test_autodiff/test_functions/test_trigonometric.py]: This use of `backward` will change, once you update it to the proper signature

**Review by VIVelev — CHANGES_REQUESTED** [review]
Mainly:
  - *DO NOT* use nujo Functions in other nujo Functions
  - Fix the `backward` method signature

Otherwise, well done.

File: nujo/autodiff/_functions/_trigonometric.py (hunk 2 above)
Comment (ManiacMaxo) [reply nujo/autodiff/_functions/_trigonometric.py]: Do I need to add them when they aren't used?

Comment (VIVelev) [reply nujo/autodiff/_functions/_trigonometric.py]: yep, since the API needs to be the same for all the functions

**Comment by ManiacMaxo** [comment]
@VIVelev Addressed the comments

**Pull request merged**